src/wayland-eglsurface.c

```diff
@@ -1135,40 +1144,33 @@ wlEglSurfaceCheckReleasePoints(WlEglDisplay *display, WlEglSurface *surface)
     EGLDisplay          dpy         = display->devDpy->eglDisplay;
     EGLSyncKHR          releaseSync = EGL_NO_SYNC_KHR;
     WlEglStreamImage   *image = NULL;
-    uint32_t           *syncobjs;
-    uint64_t           *syncPoints;
-    uint32_t            i, firstSignaled, numSyncPoints = 0;
+    WlEglStreamImage   *streamImages[MAX_IMAGES];
+    uint32_t            syncobjs[MAX_IMAGES];
+    uint64_t            syncPoints[MAX_IMAGES];
+    uint32_t            firstSignaled, numSyncPoints = 0;
     int64_t             timeout;
     EGLBoolean          ret = EGL_FALSE;
 
     if (!surface->wlSyncobjSurf) {
         return EGL_TRUE;
     }
 
-    syncobjs = calloc(surface->ctx.numStreamImages, sizeof(uint32_t));
-    syncPoints = calloc(surface->ctx.numStreamImages, sizeof(uint64_t));
-    if (!syncobjs || !syncPoints) {
-        return EGL_FALSE;
-    }
-
-    for (i = 0; i < surface->ctx.numStreamImages; i++) {
-        pthread_mutex_lock(&surface->ctx.streamImages[i]->mutex);
-    }
+    pthread_mutex_lock(&surface->ctx.streamImagesMutex);
 
     /* record each release point we are waiting on */
-    for (i = 0; i < surface->ctx.numStreamImages; i++) {
-        syncobjs[i] = surface->ctx.streamImages[i]->drmSyncobjHandle;
+    wl_list_for_each(image, &surface->ctx.streamImages, link) {
+        if (image->releasePending) {
+            streamImages[numSyncPoints] = image;
+            syncobjs[numSyncPoints] = image->drmSyncobjHandle;
+            syncPoints[numSyncPoints] = image->releasePoint;
 
-        if (surface->ctx.streamImages[i]->releasePending) {
-            syncPoints[i] = surface->ctx.streamImages[i]->releasePoint;
             numSyncPoints++;
-        } else {
-            /*
-             * Use a bogus point for acquired images so we can keep our indices
-             * the same. This won't affect anything since it will never have a fence
-             * appear.
-             */
-            syncPoints[i] = UINT64_MAX;
+
+            if (numSyncPoints >= MAX_IMAGES) {
```

Comment:
This is incorrect if we have exactly MAX_IMAGES elements, because we'll have already incremented numSyncPoints to MAX_IMAGES.

Instead, this check needs to go before the assignment and increment above.

Or alternatively, we could have an extra loop to count the number of elements we'll need and then call `alloca` to allocate the arrays for them.

Reply:
numSyncPoints is incremented to n right after the elements at the index n-1 is assigned. We check for this condition right after the increment and before the loop which we assign the elements at the index n. So as long as the initial value of numSyncPoints is less than MAX_IMAGES, which currently is, this should work. Though I believe moving the check above the assignment as you suggested would be less error-prone. We wouldn't be relying on an assumption of the initial value of numSyncPoints.